Google Pay integration.

dbrant opened the pull request:
https://phabricator.wikimedia.org/T362698

cooltey commented on app/src/extra/res/layout/activity_donate.xml:
```diff
@@ -0,0 +1,167 @@
+
+            </com.google.android.material.textfield.TextInputLayout>
+
+            <CheckBox
```
Please apply a proper text color to prevent the text is not visible in dark mode.

dbrant replied:
Ah, can't believe I missed these!

cooltey commented on app/src/extra/res/layout/activity_donate.xml:
```diff
@@ -0,0 +1,167 @@
+                android:text="@string/donate_gpay_check_transaction_fee"
+                />
+
+            <CheckBox
```
Same

cooltey commented on app/src/extra/res/layout/activity_donate.xml:
```diff
@@ -0,0 +1,167 @@
+                android:text="@string/donate_gpay_check_recurring"
+                />
+
+            <CheckBox
```
Same

cooltey commented on app/src/extra/res/layout/activity_donate.xml:
```diff
@@ -0,0 +1,167 @@
+
+            </androidx.constraintlayout.widget.ConstraintLayout>
+
+            <com.google.android.material.textfield.TextInputLayout
```
The dollar sign seems to be to slightly invisible for dark theme.

dbrant commented on app/src/extra/java/org/wikipedia/donate/GooglePayViewModel.kt:
```diff
@@ -0,0 +1,150 @@
+            if (Prefs.paymentMethodsMerchantId.isEmpty() || Prefs.paymentMethodsGatewayId.isEmpty()) {
+                uiState.value = NoPaymentMethod()
+            } else {
+                uiState.value = Resource.Success(donationConfig!!)
```
I'm actually not sure this would be safer...  If `donationConfig` is null, it will throw a NullPointerException, which will be caught by the `CoroutineExceptionHandler` at the top, and will automatically provide a `Resource.Error`.

cooltey reviewed (changes requested):
Two questions:
1. When I changed the country code to TW (Taiwan), it showed the currency correctly, but it did not show the suggested amounts, is that expected? 
<img src="https://github.com/wikimedia/apps-android-wikipedia/assets/2435576/082c1bd7-8dad-4ad8-b6b4-3ae26953b628" width="450" />
<img src="https://github.com/wikimedia/apps-android-wikipedia/assets/2435576/2e14f47f-b2b1-42e6-9bb2-bf1b52b7d3d1" width="450" />

2. It looks like the input field allows the user to add multiple zeros to the beginning of the amount, and it does not affect the final amount. Just wonder if we need to remove the unnecessary zero automatically.
<img src="https://github.com/wikimedia/apps-android-wikipedia/assets/2435576/e7cf1576-5a6d-4e1c-9c02-33c1bf57bdde" width="450" />

3. When processing the payment, the activity title still showed the "Select amount" title with an empty content. Do we need to change the activity title to "Processing payment" or other proper title?

dbrant commented:
@cooltey Excellent observations!

> 1. When I changed the country code to TW (Taiwan), it showed the currency correctly, but it did not show the suggested amounts, is that expected?

This actually might be a bug in the fundraising API -- it's not giving us the preset amounts for the TWD currency.
I'll make a note for the FR-tech team, and I updated the PR to bounce out of the GPay workflow (fall back to the browser) if the currency presets are not available. Thanks!

> 2. It looks like the input field allows the user to add multiple zeros to the beginning of the amount, and it does not affect the final amount. Just wonder if we need to remove the unnecessary zero automatically.

Ehhh I think that's an acceptable behavior. As long as the final amount is parsed correctly, why not let people enter 000001 when they really mean 1.

> 3. When processing the payment, the activity title still showed the "Select amount" title with an empty content. Do we need to change the activity title to "Processing payment" or other proper title?

¯\\\_(ツ)\_/¯  this is how it was designed, and it's probably perfectly good for a first iteration. We will surely make refinements as we go along, and I'll make an additional note to discuss this step at the next prioritization.

cooltey commented:
> @cooltey Excellent observations!
> 
> > 1. When I changed the country code to TW (Taiwan), it showed the currency correctly, but it did not show the suggested amounts, is that expected?
> 
> This actually might be a bug in the fundraising API -- it's not giving us the preset amounts for the TWD currency. I'll make a note for the FR-tech team, and I updated the PR to bounce out of the GPay workflow (fall back to the browser) if the currency presets are not available. Thanks!

Great! Thanks!

> 
> > 2. It looks like the input field allows the user to add multiple zeros to the beginning of the amount, and it does not affect the final amount. Just wonder if we need to remove the unnecessary zero automatically.
> 
> Ehhh I think that's an acceptable behavior. As long as the final amount is parsed correctly, why not let people enter 000001 when they really mean 1.

That makes sense to me.

> 
> > 3. When processing the payment, the activity title still showed the "Select amount" title with an empty content. Do we need to change the activity title to "Processing payment" or other proper title?
> 
> ¯\_(ツ)_/¯ this is how it was designed, and it's probably perfectly good for a first iteration. We will surely make refinements as we go along, and I'll make an additional note to discuss this step at the next prioritization.

Got it!


---

I have one final question before merging:

From the article donate campaign dialog, when the user selects "Donate now", the GPay bottomsheet pops up. If the user accidentally closes the bottomsheet by tapping on the dimmed side, the entire donation flow will still be considered as "Completed".

Do we need to update the preference logic to make sure the user clicks one of the options in the GPay bottomsheet instead of saving the preference when tapping on the the dialog "Donate now"?
